### PR TITLE
TCP_NOTSENT_LOWAT socket option support

### PR DESCRIPTION
**Pull request by Scottmitch**
Motiviation:
Linux provides the TCP_NOTSENT_LOWAT socket option. This can be used to control how much unsent data is queued in the tcp kernel buffers. This can be important when application level protocols (SPDY, HTTP/2) have their own priority mechanism and don't want data queued in the kernel.

Modifications:
- The epoll module will have an additional socket option TCP_NOTSENT_LOWAT
- There will be JNI methods to control the underlying linux socket option mechanism

Result:
Linux EPOLL module exposes the TCP_NOTSENT_LOWAT socket option.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer @ejona86 - FYI

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 1]
```diff
@@ -1182,7 +1182,11 @@ JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setKeepAlive(JNIEnv* e
 }
 
 JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setTcpCork(JNIEnv* env, jclass clazz, jint fd, jint optval) {
-    setOption(env, fd, SOL_TCP, TCP_CORK, &optval, sizeof(optval));
+    setOption(env, fd, IPPROTO_TCP, TCP_CORK, &optval, sizeof(optval));
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 2]
```diff
@@ -1206,15 +1210,15 @@ JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setBroadcast(JNIEnv* e
 }
 
 JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setTcpKeepIdle(JNIEnv* env, jclass clazz, jint fd, jint optval) {
-    setOption(env, fd, SOL_TCP, TCP_KEEPIDLE, &optval, sizeof(optval));
+    setOption(env, fd, IPPROTO_TCP, TCP_KEEPIDLE, &optval, sizeof(optval));
 }
 
 JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setTcpKeepIntvl(JNIEnv* env, jclass clazz, jint fd, jint optval) {
-    setOption(env, fd, SOL_TCP, TCP_KEEPINTVL, &optval, sizeof(optval));
+    setOption(env, fd, IPPROTO_TCP, TCP_KEEPINTVL, &optval, sizeof(optval));
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 3]
```diff
@@ -1206,15 +1210,15 @@ JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_setBroadcast(JNIEnv* e
...
 }
 
 JNIEXPORT void Java_io_netty_channel_epoll_Native_setTcpKeepCnt(JNIEnv* env, jclass clazz, jint fd, jint optval) {
-    setOption(env, fd, SOL_TCP, TCP_KEEPCNT, &optval, sizeof(optval));
+    setOption(env, fd, IPPROTO_TCP, TCP_KEEPCNT, &optval, sizeof(optval));
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 4]
```diff
@@ -1259,7 +1263,15 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getSendBufferSize(JNIE
 
 JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_isTcpCork(JNIEnv* env, jclass clazz, jint fd) {
     int optval;
-    if (getOption(env, fd, SOL_TCP, TCP_CORK, &optval, sizeof(optval)) == -1) {
+    if (getOption(env, fd, IPPROTO_TCP, TCP_CORK, &optval, sizeof(optval)) == -1) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 5]
```diff
@@ -1303,7 +1315,7 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_isBroadcast(JNIEnv* en
 
 JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepIdle(JNIEnv* env, jclass clazz, jint fd) {
     int optval;
-    if (getOption(env, fd, SOL_TCP, TCP_KEEPIDLE, &optval, sizeof(optval)) == -1) {
+    if (getOption(env, fd, IPPROTO_TCP, TCP_KEEPIDLE, &optval, sizeof(optval)) == -1) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 6]
```diff
@@ -1311,7 +1323,7 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepIdle(JNIEnv*
 
 JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepIntvl(JNIEnv* env, jclass clazz, jint fd) {
     int optval;
-    if (getOption(env, fd, SOL_TCP, TCP_KEEPINTVL, &optval, sizeof(optval)) == -1) {
+    if (getOption(env, fd, IPPROTO_TCP, TCP_KEEPINTVL, &optval, sizeof(optval)) == -1) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 7]
```diff
@@ -1319,7 +1331,7 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepIntvl(JNIEnv
 
 JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepCnt(JNIEnv* env, jclass clazz, jint fd) {
      int optval;
-     if (getOption(env, fd, SOL_TCP, TCP_KEEPCNT, &optval, sizeof(optval)) == -1) {
+     if (getOption(env, fd, IPPROTO_TCP, TCP_KEEPCNT, &optval, sizeof(optval)) == -1) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c [hunk 8]
```diff
@@ -1327,7 +1339,7 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_getTcpKeepCnt(JNIEnv*
 
 JNIEXPORT void JNICALL Java_io_netty_channel_epoll_Native_tcpInfo0(JNIEnv* env, jclass clazz, jint fd, jintArray array) {
      struct tcp_info tcp_info;
-     if (getOption(env, fd, SOL_TCP, TCP_INFO, &tcp_info, sizeof(tcp_info)) == -1) {
+     if (getOption(env, fd, IPPROTO_TCP, TCP_INFO, &tcp_info, sizeof(tcp_info)) == -1) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: why was this changed ? Was it incorrect?

**Comment by normanmaurer** [comment]
@Scottmitch just a few questions

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c (hunk 1 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: `SOL_TCP` is an isolated define from `netinet/tcp.h`.  `IPPROTO_TCP` is defined with all other "common" protocols in `netinet/in.h` and is the recommended enumeration to use for TCP [1].  These definitions are both functionally equivalent and resolve to `6` however I believe `IPPROTO_TCP` is more correct to use.  In addition `SOL_TCP` is no longer visible now that we don't include `netinet/tcp.h`.

[1] http://linux.die.net/man/7/tcp

Comment (Scottmitch) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: @normanmaurer ^

**Comment by Scottmitch** [comment]
@normanmaurer - I just responded to you comment on the first occurrence. I can address the others if necessary, but I think we are OK as is.

File: transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c (hunk 1 above)
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c]: I see... Ok go ahead and merge then :)

**Comment by Scottmitch** [comment]
Cherry-picked 4.0 (a17e1c8) 4.1 (5121fba) master (8caa584)
